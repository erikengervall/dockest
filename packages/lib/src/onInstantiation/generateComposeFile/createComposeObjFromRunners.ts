import { ComposeService, DependsOn, ComposeFile } from '../../runners/@types'
import { DockestConfig } from '../../index'

const { keys } = Object

const getDepComposeServices = (dependsOn: DependsOn) =>
  dependsOn.reduce(
    (composeServices: { [key: string]: ComposeService }, { runnerConfig: { service }, getComposeService }) => ({
      ...composeServices,
      [service]: getComposeService(),
    }),
    {},
  )

export default (config: DockestConfig, dockerComposeFileVersion: string) => {
  const composeObj: ComposeFile = {
    version: dockerComposeFileVersion,
    services: {},
  }

  config.runners.forEach(runner => {
    const {
      runnerConfig: { service, dependsOn },
      getComposeService,
    } = runner

    const composeService = getComposeService()
    const depComposeServices = getDepComposeServices(dependsOn)

    composeObj.services = {
      ...composeObj.services,
      [service]: { ...composeService },
      ...depComposeServices,
    }
  })

  const servicesWithNetworks = keys(composeObj.services).filter(
    service => composeObj.services[service].networks !== undefined,
  )
  if (servicesWithNetworks.length > 0) {
    composeObj.networks = servicesWithNetworks.reduce(
      (networks, service) => ({
        ...networks,
        ...composeObj.services[service].networks,
      }),
      {},
    )
  }

  const servicesWithVolumes = keys(composeObj.services).filter(
    service => composeObj.services[service].volumes !== undefined,
  )
  if (servicesWithVolumes.length > 0) {
    composeObj.volumes = servicesWithVolumes.reduce(
      (volumes, service) => ({
        ...volumes,
        ...composeObj.services[service].volumes,
      }),
      {},
    )
  }

  return composeObj
}
