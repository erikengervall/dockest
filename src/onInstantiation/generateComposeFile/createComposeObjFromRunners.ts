import execa from 'execa'
import { ComposeService, DependsOn, ComposeFile } from '../../runners/@types'
import { DockestConfig } from '../../index'

const { keys } = Object

const getDepComposeServices = (dependsOn: DependsOn) =>
  dependsOn.reduce(
    (composeServices: { [key: string]: ComposeService }, { runnerConfig: { service }, getComposeService }) => ({
      ...composeServices,
      [service]: getComposeService(),
    }),
    {},
  )

export default (config: DockestConfig, dockerComposeFileVersion: string) => {
  const composeObj: ComposeFile = {
    version: dockerComposeFileVersion,
    services: {},
  }

  const extra_hosts: string[] = []

  if (!config.$.isInsideDockerContainer) {
    if (process.platform === 'linux') {
      const command = `ip -4 addr show docker0 | grep -oP '(?<=inet\s)\d+(\.\d+){3}'`
      const result = execa.sync(command, { reject: false })
      extra_hosts.push(`host.docker.internal:${result.stdout}`)
    }
  }

  config.runners.forEach(runner => {
    const {
      runnerConfig: { service, dependsOn },
      getComposeService,
    } = runner

    const composeService = getComposeService()
    const depComposeServices = getDepComposeServices(dependsOn)

    composeObj.services = {
      ...composeObj.services,
      [service]: { ...composeService, extra_hosts },
      ...depComposeServices,
    }
  })

  const servicesWithNetworks = keys(composeObj.services).filter(
    service => composeObj.services[service].networks !== undefined,
  )
  if (servicesWithNetworks.length > 0) {
    composeObj.networks = servicesWithNetworks.reduce(
      (networks, service) => ({
        ...networks,
        ...composeObj.services[service].networks,
      }),
      {},
    )
  }

  const servicesWithVolumes = keys(composeObj.services).filter(
    service => composeObj.services[service].volumes !== undefined,
  )
  if (servicesWithVolumes.length > 0) {
    composeObj.volumes = servicesWithVolumes.reduce(
      (volumes, service) => ({
        ...volumes,
        ...composeObj.services[service].volumes,
      }),
      {},
    )
  }

  return composeObj
}
