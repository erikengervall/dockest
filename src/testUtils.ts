import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
