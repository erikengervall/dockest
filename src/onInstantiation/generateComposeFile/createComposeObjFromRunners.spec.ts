import createComposeObjFromRunners from './createComposeObjFromRunners'
import testUtils from '../../testUtils'

const { createDockestConfig, initializedRunners } = testUtils({})

describe('createComposeObjFromRunners', () => {
  it('should create composeObj from single runner', () => {
    const dockestConfig = createDockestConfig({
      runners: [initializedRunners.postgresRunner],
    })

    const composeObj = createComposeObjFromRunners(dockestConfig)

    expect(composeObj).toMatchSnapshot()
  })

  it('should create composeObj from all initialized runners', () => {
    const dockestConfig = createDockestConfig({
      runners: Object.values(initializedRunners),
    })

    const composeObj = createComposeObjFromRunners(dockestConfig)

    expect(composeObj).toMatchSnapshot()
  })
})
