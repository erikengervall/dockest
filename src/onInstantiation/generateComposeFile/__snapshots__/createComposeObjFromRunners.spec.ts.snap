// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComposeObjFromRunners should create composeObj from all initialized runners 1`] = `
Object {
  "services": Object {
    "general": Object {
      "image": "general/image:123",
      "ports": Array [],
    },
    "kafka": Object {
      "depends_on": Array [
        "zookeeper",
      ],
      "environment": Object {
        "KAFKA_ADVERTISED_LISTENERS": "PLAINTEXT://kafka:29092,PLAINTEXT_HOST://localhost:9092",
        "KAFKA_AUTO_CREATE_TOPICS_ENABLE": "true",
        "KAFKA_BROKER_ID": 1,
        "KAFKA_LISTENER_SECURITY_PROTOCOL_MAP": "PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT",
        "KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR": 1,
        "KAFKA_ZOOKEEPER_CONNECT": "zookeeper:2181",
      },
      "image": "kafka/image:123",
      "ports": Array [
        Object {
          "published": 9092,
          "target": 9092,
        },
      ],
    },
    "postgres": Object {
      "environment": Object {
        "POSTGRES_DB": "_",
        "POSTGRES_PASSWORD": "_",
        "POSTGRES_USER": "_",
      },
      "image": "postgres/image:123",
      "ports": Array [
        Object {
          "published": 5432,
          "target": 5432,
        },
      ],
    },
    "redis": Object {
      "image": "redis/image:123",
      "ports": Array [
        Object {
          "published": 6379,
          "target": 6379,
        },
      ],
    },
    "zookeeper": Object {
      "environment": Object {
        "ZOOKEEPER_CLIENT_PORT": 2181,
      },
      "image": "zookeeper/image:123",
      "ports": Array [
        Object {
          "published": 2181,
          "target": 2181,
        },
      ],
    },
  },
  "version": "3.0",
}
`;

exports[`createComposeObjFromRunners should create composeObj from single runner 1`] = `
Object {
  "services": Object {
    "postgres": Object {
      "environment": Object {
        "POSTGRES_DB": "_",
        "POSTGRES_PASSWORD": "_",
        "POSTGRES_USER": "_",
      },
      "image": "postgres/image:123",
      "ports": Array [
        Object {
          "published": 5432,
          "target": 5432,
        },
      ],
    },
  },
  "version": "3.0",
}
`;
